### PR TITLE
GAPI Fluid SIMD:Add support of new several types for the Merge3.

### PR DESCRIPTION
**Pull request by anna-khakimova**
- Support of the new several types was added.
- Fixes for the Split/Merge and ConvertTo issues.

<cut/>

```
force_builders=Linux AVX2,Linux32,Win32,Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.2a
build_gapi_standalone:Win64=ade-0.1.2a
Xbuild_gapi_standalone:Mac=ade-0.1.2a
build_gapi_standalone:Linux x64 Debug=ade-0.1.2a

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.2a

Xbuild_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.2.0

buildworker:Custom Win=windows-3

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
Xtest_opencl:Custom=OFF
Xtest_bigdata:Custom=1
Xtest_filter:Custom=*

CPU_BASELINE:Custom Win=AVX512_SKX
CPU_BASELINE:Custom=SSE4_2
```

### CODE REVIEW

File: modules/gapi/src/backends/fluid/gfluidutils.hpp [hunk 1]
```diff
@@ -86,6 +86,25 @@ using cv::gapi::own::rintd;
         return;                                            \
     }
 
+#define MERGE3_(T, OP, ...)                                \
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidutils.hpp:89]: Cannot find place where is it used, could you point me out?

Comment (sivanov-work) [reply modules/gapi/src/backends/fluid/gfluidutils.hpp:89]: discussed offline:
to move out repeated conditions into beginning of function to simplify condition calculation complexity in actual type dispatching routing

**Review by sivanov-work — APPROVED** [review]
LGTM

File: modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp [hunk 2]
```diff
@@ -252,6 +252,7 @@ INSTANTIATE_TEST_CASE_P(Split4PerfTestCPU, Split4PerfTest,
 INSTANTIATE_TEST_CASE_P(Merge3PerfTestCPU, Merge3PerfTest,
     Combine(Values(AbsExact().to_compare_f()),
             Values(szSmall128, szVGA, sz720p, sz1080p),
+            Values(CV_8U),
```
Comment (dmatveev) [inline modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp:255]: Should other types be listed here?

File: modules/gapi/src/backends/fluid/gfluidutils.hpp [hunk 3]
```diff
@@ -86,6 +86,23 @@ using cv::gapi::own::rintd;
...
+        GAPI_DbgAssert(dst.length() == src3.length());     \
+                                                           \
+        GAPI_DbgAssert(1 == src1.meta().chan);             \
+        GAPI_DbgAssert(1 == src2.meta().chan);             \
+        GAPI_DbgAssert(1 == src3.meta().chan);             \
+        GAPI_DbgAssert(3 == dst.meta().chan);              \
+                                                           \
+        OP<T>(__VA_ARGS__);                                \
+        return;                                            \
+    }
```
Comment (dmatveev) [inline modules/gapi/src/backends/fluid/gfluidutils.hpp:104]: Macro function bodies considered harmful, is there any reasonable issues with rewriting it to templates?

**Comment by asmorkalov** [comment]
@anna-khakimova @dmatveev Do you plan to finish the PR in meantime?

**Comment by asmorkalov** [comment]
@dmatveev @TolyaTalamanov Friendly reminder.

**Comment by asmorkalov** [comment]
@dmatveev @TolyaTalamanov Friendly reminder.

**Comment by dmatveev** [comment]
Will have a look at this one too.

**Comment by dmatveev** [comment]
@rgarnov can you please have a look?

**Review by dmatveev — APPROVED** [review]
Tested locally, these changes don't bring any regressions on my end but so far extend Fluid's merge3 with extra data types.

Can be merged IMO
```
Geometric mean (ms)

                                               Name of Test                                                  4x   anna     anna
                                                                                                                  simd     simd
                                                                                                                            vs
                                                                                                                            4x
                                                                                                                        (x-factor)
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 128x128, 8UC1, { gapi.kernel_package })      -   0.016     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 128x128, 16UC1, { gapi.kernel_package })     -   0.019     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 128x128, 16SC1, { gapi.kernel_package })     -   0.019     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 128x128, 32FC1, { gapi.kernel_package })     -   0.025     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 128x128, { gapi.kernel_package })          0.016   -       -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 640x480, 8UC1, { gapi.kernel_package })      -   0.099     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 640x480, 16UC1, { gapi.kernel_package })     -   0.175     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 640x480, 16SC1, { gapi.kernel_package })     -   0.178     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 640x480, 32FC1, { gapi.kernel_package })     -   0.347     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 640x480, { gapi.kernel_package })          0.103   -       -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1280x720, 8UC1, { gapi.kernel_package })     -   0.255     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1280x720, 16UC1, { gapi.kernel_package })    -   0.621     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1280x720, 16SC1, { gapi.kernel_package })    -   0.613     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1280x720, 32FC1, { gapi.kernel_package })    -   1.729     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1280x720, { gapi.kernel_package })         0.258   -       -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1920x1080, 8UC1, { gapi.kernel_package })    -   0.781     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1920x1080, 16UC1, { gapi.kernel_package })   -   1.985     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1920x1080, 16SC1, { gapi.kernel_package })   -   1.975     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1920x1080, 32FC1, { gapi.kernel_package })   -   4.211     -
TestPerformance::Merge3PerfTestFluid/Merge3PerfTest::(compare_f, 1920x1080, { gapi.kernel_package })        0.744   -       -
TestPerformance::Merge4PerfTestFluid/Merge4PerfTest::(compare_f, 128x128, { gapi.kernel_package })          0.019 0.020    0.96
TestPerformance::Merge4PerfTestFluid/Merge4PerfTest::(compare_f, 640x480, { gapi.kernel_package })          0.127 0.129    0.98
TestPerformance::Merge4PerfTestFluid/Merge4PerfTest::(compare_f, 1280x720, { gapi.kernel_package })         0.361 0.348    1.04
TestPerformance::Merge4PerfTestFluid/Merge4PerfTest::(compare_f, 1920x1080, { gapi.kernel_package })        1.228 1.221    1.01
```

**Comment by asmorkalov** [comment]
@opencv-alalek Could you take a look on CI failures. It looks like some update is required.

**Comment by dmatveev** [comment]
Yep the main build runs ok:

> -- ADE: Downloading v0.1.2a.zip from https://github.com/opencv/ade/archive/v0.1.2a.zip

But there's a separate ADE step where another (older) version is mentioned:

> CMake Error: The source directory "/build/precommit_linux64_no_opt/build/3rdparty/ade/ade-0.1.1f" does not exist.

**Comment by opencv-alalek** [comment]
> 0.1.1f

Because it is mentioned in CI configuration section of **PR's description**.

**Comment by dmatveev** [comment]
> Because it is mentioned in CI configuration section of PR's description.

Cool! Didn't know it still works given that there's Github Actions :)

**Pull request merged**